packages/fxa-content-server/app/tests/spec/lib/email-domain-validator.js

```diff
@@ -0,0 +1,170 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import $ from 'jquery';
+import { assert } from 'chai';
+import AuthErrors from 'lib/auth-errors';
+import checkEmailDomain from 'lib/email-domain-validator';
+import p from 'lib/promise';
+import sinon from 'sinon';
+import Tooltip from 'views/tooltip';
+import TopEmailDomains from 'lib/top-email-domains';
+import xhr from 'lib/xhr';
+
+// This is from the email-domain-validator module
+const REJECTION_CODE = '__INVALID_EMAIL_DOMAIN__';
+
+const TEST_DOMAIN = 'abc.xyz';
+const TOOLTIP_SELECTOR = '.tooltip-suggest';
+let $el;
+let view;
+
+describe('lib/email-domain-validator', () => {
+  beforeEach(() => {
+    view = {
+      logEvent: sinon.stub(),
+    };
+
+    $('body').append(
+      '<div class="input-row test-input"><input type="email" id="email" name="email"/></div>'
+    );
+    $el = $('.test-input');
+    $el.val(`quuz@${TEST_DOMAIN}`);
+  });
+
+  afterEach(() => {
+    $('.test-input').remove();
+  });
+
+  it('should check the list of top domains', () => {
+    sinon.stub(TopEmailDomains, 'has').returns(true);
+    sinon.spy(xhr, 'ajax');
+    return checkEmailDomain($el, view).then(() => {
+      assert.isTrue(TopEmailDomains.has.calledOnceWith(TEST_DOMAIN));
+      assert.isFalse(xhr.ajax.called);
+      xhr.ajax.restore();
+      TopEmailDomains.has.restore();
+    });
+  });
+
+  it('should check if the domain has previously failed validation and reject', () => {
+    $el.data('previousDomain', TEST_DOMAIN);
+    $el.data('previousDomainResult', 'none');
+    sinon.spy($el, 'data');
+    return checkEmailDomain($el, view)
+      .then(() => assert.fail('The promise should have rejected'))
+      .catch(() => {
+        assert.isTrue($el.data.calledWith('previousDomain'));
+        assert.isTrue($el.data.calledWith('previousDomainResult'));
+      });
+  });
+
+  it('should check if the domain has been previously resolved with an A record and resolve', () => {
+    $el.data('previousDomain', TEST_DOMAIN);
+    $el.data('previousDomainResult', 'A');
+    sinon.spy($el, 'data');
+    return checkEmailDomain($el, view).then(() => {
+      assert.isTrue($el.data.calledWith('previousDomain'));
+      assert.isTrue($el.data.calledWith('previousDomainResult'));
+    });
+  });
+
+  it('should resolve on a successful MX record lookup', () => {
+    sinon.stub(xhr, 'ajax').returns(Promise.resolve({ result: 'MX' }));
```

Comment:
TIL about `sinon.stub().returns` and `sinon.stub().resolves`, which looks like you could use the latter here?

Reply:
Huh, it did not occur to me whatsoever, but looks like I could.  I'm not sure what we'd gain from it though?